torch_ttnn/passes/lowering/to_tt_pass.py

```diff
@@ -852,6 +876,64 @@ def rewrite_node(node):
                 output_4d = g.call_function(ttnn.moreh_cumsum, (input_4d, dim), kwargs)
                 return g.call_function(ttnn.reshape, (output_4d, input_shape))
 
+            if node.target == torch.ops.aten.max_pool2d_with_indices.default:
+                params = map_args_to_kwargs(
+                    args,
+                    (
+                        (0, "input_tensor"),
+                        (1, "kernel_size"),
+                        (2, "stride"),
+                        (3, "padding"),
+                        (4, "dilation"),
+                        (5, "ceil_mode"),
+                    ),
+                )
+                input_tensor = params["input_tensor"]
+                kernel_size = params["kernel_size"]
+                batch_size, in_c, in_h, in_w = input_tensor.meta["val"].size()
+                stride = params.get("stride", kernel_size)
+                padding = params.get("padding", (0, 0))
+                dilation = params.get("dilation", (1, 1))
+                ceil_mode = params.get("ceil_mode", False)
+                # Assume the element size is bfloat16
+                volume = (batch_size * in_c * in_h * in_w) * 2
+                if (
+                    # TODO(tt-metal#14976): ceil mode isn't supported yet
+                    ceil_mode
+                    # TODO(#385): OOM
+                    or volume > 16 * 1024 * 1024
+                    # TODO(tt-metal#13901): Non-pow-of-2 channel isn't supported yet
+                    or (in_c & (in_c - 1)) != 0
+                    # TODO(#419): Currently fails with in_c < 16
+                    or in_c < 16
+                    # TODO(tt-metal#12099): Currently it doesn't return indices. Convert if only the value is used
+                    or not is_getitem_0_only_user(node)
+                ):
+                    return None
+
+                input_tensor = insert_nchw_to_nhwc(g, input_tensor)
+                # TODO(#423): reshape can be removed if (N, H, W, C) is supported
+                input_tensor = g.call_function(ttnn.reshape, (input_tensor, (1, 1, batch_size * in_h * in_w, in_c)))
+                # TODO(#418): max_pool2d fails with input tensors in tile layout for some shapes
+                input_tensor = g.call_function(ttnn.to_layout, (input_tensor, TtnnRowMajorLayout()))
```

Comment:
I assume if we first change layout and then do the rest of the processing we can optimize layout changes out.

Reply:
Yes, currently there will be many redundant `to_layout` (and even `from_device`, `to_device`). The idea in my mind is to not specially handle them here but let the following pass to optimize them out. Currently we already have `eliminate_coreops_pass.py` to eliminate certain `from_device - to_device` pairs and we can improve it to remove more redundancy.

A more complicated optimization pass can also be done in the future to hoist or sink to_layout/reshape across tt ops (if they can still run), which will reduce the number of layout change/reshape between ops. For example: the `(1, 1, NHW, C)` output from `max_pool2d` doesn't need to be permuted or reshaped immediately if the following is an unary pointwise (e.g. relu)